### ISSUE TITLE
Updates

### ISSUE THREAD

**Issue by turtlepod** [Astoundify/wp-job-manager-companies#19]
So, I'll fix several issue in this plugin, and will report in this thread.

**Comment by turtlepod** [comment]
#### Day 1

- I create a new branch https://github.com/Astoundify/wp-job-manager-companies/tree/david
- Exclude filled listing & expired #14 
- Create a new group "0-9" to hold all companies that don't start with "A" - "Z" probably will solve #15 #10 
- URL seem to be working well for non english language too.
- Fix output buffering in shortcode output

I also added reply in other unresolved thread:
- Canonical URL #18 
- Optimize for Large Amounts #11

**Comment by adampickering** [comment]
Great so far. I was wondering if you could look over all of these topics to see if there is anything we can do to improve the plugins code https://wordpress.org/support/plugin/wp-job-manager-companies

**Comment by turtlepod** [comment]
I just did. No new issue, basically the same issue as we have here:
- language support.
- change the grouping (a-z)
- add meta data to companies (logo, desc)
- blank page (I actually fix this, by adding index.php as fallback template)

so, there are two main way, keep continue this plugin as is (company as custom field), and we will never have feature like changing slug, adding logo, (since its not possible) and let user know the limitation of this plugin.
note: it's perfectly fine, but we have to understand the limitation between post meta vs taxonomy.

or move to custom taxonomy, which we can do all that out of the box,
but we need to figure out how to transfer this company data to taxonomy term using upgrade routine, and keep this "sync" with post meta, so when they disable the plugin this company data is still there (probably not needed, it probably will be one way street)

**Comment by spencerfinnell** [comment]
The idea of using a custom taxonomy is definitely cool. The syncing part is definitely the annoying part. We would need to keep them in sync though so the plugin could be disabled like you said. Though that shouldn't be too bad by just watching `save_post` and `save_term` (or whatever it is). 

I like that solution as well since it allows companies to be created without having a job listing created, which I've seen requested quite a few times in support. 

WP Job Manager also saves to user meta so we would probably just ignore that? https://github.com/Automattic/WP-Job-Manager/blob/master/includes/forms/class-wp-job-manager-form-submit-job.php#L643

**Comment by turtlepod** [comment]
It actually will be a really complex plugin too.
- company tagline
- featured image (company logo)
- company video
- company website
- company twitter
- maybe application email/url

all need to be transferred to term meta, I don't think "sync" is worth it.
I'm not even sure we can do all the transition from post meta to term taxonomy smoothly. 
(100 posts about a company can have 100 different logo, tagline, etc. which one too choose to save in term meta data? which one is real?)
and how about the job form submit/edit system, we need to filter all, remove all this fields, and add our own, and sync.

Also possible support question "previously i can add different logo to the same company, because i don't use logo as logo, but now i can no longer do this..."

in short, it's going to be a really big plugin (i don't know if it's worth it).

**Comment by spencerfinnell** [comment]
What if we just used the Company Name as a term in the taxonomy and only tracked that? When a listing is submitted we just treat it like we are assigning the listing to any other taxonomy. The UI could be hidden on the frontend and backend (on the job listing page). 

This would give us real URLs, better queries, etc, but without having to worry about the rest of the meta. The way the company profile page is implemented in Jobify currently doesn't account for any differences in the other information anyway. It uses the first result and bases the rest of the company information on that. So even if they are different between individual jobs with the same company name it has never shown that way.

**Comment by turtlepod** [comment]
that is good idea.
plus if we have all this data in the term, it will be complicated to associate user with each term (company).
a user edit the company logo, and it should not effect other user jobs with the same company.

But It still is very tricky. For example, if a user change the company name (a typo maybe), what should we do? (a lot if little things like this, and we need to decide how to handle this).

**Comment by spencerfinnell** [comment]
When a listing is saved we just set the company name as a term. If it doesn't exist it will be created, or it will assign that listing to an existing term. 

This would produce the same outcome as the current plugin. If you spell one Company Name "AirBnb" and another "Air Bnb" (with a space) it would show up twice in the current plugin. 

To help admins provide consistency we could remove the _company_name field in the admin and replace it with the actual taxonomy UI on the post edit screen, but that might give too much "disconnect" (figuratively speaking) from the rest of the company information. Perhaps replace the text input then with a select field of all existing terms? Then it could stay in the same meta box. 

That just leaves the other direction: modifying an existing company name term. It would show updated in the companies archive but would not show up differently on an individual job. But I think that's okay?

**Comment by turtlepod** [comment]
if you want to go this route, we will need proper wireframe of the flow.

for example user input "Arbnb" a typo, when user correct it in front end to "AirBnB" it should check if the term "Arbnb" already have other job/post attach (more than one/other than current job), it it didn't have other job, we should simply edit the term name (and maybe slug) with new data.
Or if other job have the company "arbnb" (maybe it's not a typo afterall), we should create a new term "AirBnB" and leave "arbnb" company as is, because other job have it.

and it's a just small example of what we are facing. But I'm sure there's a lot of little things like this :)
(it's going to be fun)

**Comment by spencerfinnell** [comment]
Yeah I think that is a good example of the typo scenario and a good solution. Can you list out any other scenarios you can think of? 

To me the flow is:

1. Create a listing

2. Assign the value of _company_name to a term in `job_listing_company`. It will act just like a tag: if it exists it will assign to the current one, otherwise it will create a new one.

3. Edit _company_name in frontend/backend. Check if this is the only associated listing for the current/old value. If so update the term information with the new data. Otherwise assign to the new term.

4. If a term is edited in the admin then nothing in the associated jobs changes. The only reflection will be on the company archive.

**Comment by turtlepod** [comment]
I think that's a very good flow.
I'm not sure about other scenarios :)

**Comment by spencerfinnell** [comment]
Also need to remember redirects for the existing permalink structure.

**Comment by turtlepod** [comment]
that is correct.
good luck with the development of the plugin :)

**Comment by spencerfinnell** [comment]
Going to wait for https://github.com/Automattic/WP-Job-Manager/issues/1135

**Comment by dariushazimi** [comment]
Hello, 
Jobify latest version as of Feb 1, 2019
Just clicking on the update button on the company page, update fails.

update to the company failes. I have the defaults in the latest version of jobify with the following shortcode:
`[job_manager_companies]`
Issue: No company shows up on the home page
When I login as admin and I just click update on the company page where the shortcode` "[job_manager_companies]"` is placed. Update fails. Saying

>  "Updating failed"

. Am I missing anything?
Thanks
